Remove `spack.target` from code

alalazo opened the pull request:
The `spack.target.Target` class is a weird entity, that is just needed to:

1. Sort microarchitectures in lists deterministically
2. Being able to use microarchitectures in hashed containers

This PR removes it, and uses `archspec.cpu.Microarchitecture` directly. To sort lists, we use a proper `key=` when needed. Being able to use `Microarchitecture` objects in sets is achieved by updating the external `archspec`.

We might clean the code further, in following PRs, by introducing microarchitecture ranges in `ArchSpec`, and simplifying the constrain etc. logic.

alalazo committed:
Target: move optimization_flags method to spack.build_environment

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
Target: remove module_name from init arguments

This argument is not used anymore in any code path

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
Target: remove __lt__ and __ne__

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
Target: remove serialization methods

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
Remove spack.target altogether

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
Fix packages using Target objects directly

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
Break an import cycle on ubi8

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

pramodk commented on var/spack/repos/builtin/packages/neuron/package.py:
```diff
@@ -149,7 +150,7 @@ def cmake_args(self):
+        compilation_flags = optimization_flags(self.spec.compiler, self.spec.target)
         args.append(self.define("CMAKE_CXX_FLAGS", compilation_flags))
 
         return args
```
Just checked the neuron package, and nothing important - LGTM.

rbberger reviewed (approved):
LGTM for the lammps package.

haampie commented:
```/home/docs/checkouts/readthedocs.org/user_builds/spack/checkouts/46503/lib/spack/docs/_spack_root/lib/spack/spack/platforms/_platform.py:docstring of spack.platforms._platform.Platform.add_target:1: WARNING: py:class reference target not found: archspec.cpu.microarchitecture.Microarchitecture
```

alalazo committed:
Fix docs building

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
Use Compiler instead of CompilerSpec for a few packages

Note that these packages rely on "internal" API calls,
which are not stable. Spack injects the same flags through the
wrapper.

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo commented:
The only failure seems unrelated to the PR: https://gitlab.spack.io/spack/spack/-/jobs/12614028

haampie commented:
Adding `v0.22.3` because just archspec needs to be backported. Should've been a separate PR.

alalazo commented:
The only conflict is the sha, that can also be solved in the backport PR. We can keep archspec updates always separate in the future. For what is worth, the changes in `archspec` here are because of this PR.
